libyul/backends/evm/ControlFlowGraphBuilder.cpp

```diff
@@ -0,0 +1,472 @@
+	jump(afterSwitch);
+}
+
+void ControlFlowGraphBuilder::operator()(ForLoop const& _loop)
+{
+	ScopedSaveAndRestore scopeRestore(m_scope, m_info.scopes.at(&_loop.pre).get());
+	(*this)(_loop.pre);
+
+	std::optional<bool> constantCondition;
+	if (auto const* literalCondition = get_if<yul::Literal>(_loop.condition.get()))
+		constantCondition = valueOfLiteral(*literalCondition) != 0;
+
+	CFG::BasicBlock& loopCondition = m_graph.makeBlock();
+	CFG::BasicBlock& loopBody = m_graph.makeBlock();
+	CFG::BasicBlock& post = m_graph.makeBlock();
+	CFG::BasicBlock& afterLoop = m_graph.makeBlock();
+
+	ScopedSaveAndRestore scopedSaveAndRestore(m_forLoopInfo, ForLoopInfo{afterLoop, post});
+
+	if (constantCondition.has_value())
+	{
+		if (*constantCondition)
+		{
+			jump(loopBody);
+			(*this)(_loop.body);
+			jump(post);
+			(*this)(_loop.post);
+			jump(loopBody, true);
+		}
+		else
+			jump(afterLoop);
+	}
+	else
+	{
+		jump(loopCondition);
+		makeConditionalJump(std::visit(*this, *_loop.condition), loopBody, afterLoop);
+		m_currentBlock = &loopBody;
+		(*this)(_loop.body);
+		jump(post);
+		(*this)(_loop.post);
+		jump(loopCondition, true);
+	}
+
+	m_currentBlock = &afterLoop;
+}
+
+void ControlFlowGraphBuilder::operator()(Break const&)
+{
+	yulAssert(m_forLoopInfo.has_value(), "");
+	jump(m_forLoopInfo->afterLoop);
+	m_currentBlock = &m_graph.makeBlock();
+}
+
+void ControlFlowGraphBuilder::operator()(Continue const&)
+{
+	yulAssert(m_forLoopInfo.has_value(), "");
+	jump(m_forLoopInfo->post);
+	m_currentBlock = &m_graph.makeBlock();
+}
+
+void ControlFlowGraphBuilder::operator()(Leave const&)
+{
+	yulAssert(m_currentFunctionExit.has_value(), "");
+	m_currentBlock->exit = *m_currentFunctionExit;
+	m_currentBlock = &m_graph.makeBlock();
+}
+
+void ControlFlowGraphBuilder::operator()(FunctionDefinition const& _function)
+{
+	yulAssert(m_scope, "");
+	yulAssert(m_scope->identifiers.count(_function.name), "");
+	Scope::Function& function = std::get<Scope::Function>(m_scope->identifiers.at(_function.name));
+	m_graph.functions.emplace_back(&function);
+
+	yulAssert(m_info.scopes.at(&_function.body), "");
+	Scope* virtualFunctionScope = m_info.scopes.at(m_info.virtualBlocks.at(&_function).get()).get();
+	yulAssert(virtualFunctionScope, "");
+
+	auto&& [it, inserted] = m_graph.functionInfo.emplace(std::make_pair(&function, CFG::FunctionInfo{
+		_function.debugData,
+		function,
+		&m_graph.makeBlock(),
+		_function.parameters | ranges::views::transform([&](auto const& _param) {
+			return VariableSlot{
+				std::get<Scope::Variable>(virtualFunctionScope->identifiers.at(_param.name)),
+				_param.debugData
+			};
+		}) | ranges::to<vector>,
+		_function.returnVariables | ranges::views::transform([&](auto const& _retVar) {
+			return VariableSlot{
+				std::get<Scope::Variable>(virtualFunctionScope->identifiers.at(_retVar.name)),
+				_retVar.debugData
+			};
+		}) | ranges::to<vector>
+	}));
+	yulAssert(inserted, "");
+	CFG::FunctionInfo& info = it->second;
```

Comment:
Having `m_info` of type `AnalysisInfo` and `info` of tyle `FunctionInfo` makes this part a bit confusing.
```suggestion
	CFG::FunctionInfo& functionInfo = it->second;
```